### DISCUSSION TITLE
Crazyflies loosing poses when using Crazyswarm, Probable rootcause Crazyradios

### DISCUSSION THREAD

**Discussion by darshit-desai** [bitcraze/discussions#1016] [Q&A]
Hello Bitcraze team,

I would like to reference an issue which I faced while using Crazyswarm API (ROS1 version) and we along with the developers of crazyswarm feel that the issue can be narrowed down to Crazyradios. I am mentioning @whoenig who I discussed this issue with and ran some diagnostics about the same, here's the issue link https://github.com/USC-ACTLab/crazyswarm/discussions/769#discussion-5727725. 

Here's what my setup consists of,

24 drones communicating with 3 Crazyradio PAs over 3 different channels (80, 90 and 100), the radios have the firmware which was earmarked for broadcast operations in the Crazyswarm 1 documentation and the Crazyflies have the latest firmware released in July 2023.

The localization is done via Vicon Cameras (Depending on which lab I am testing this the number of cameras differ)

Environment (Main issue): There are two labs where I test this setup,

Lab 1: No RF shielding with 12 cameras and possibly is the most noisy environment since the lab is inside a computer science building

Lab 2: Special purpose shielding with 40 cameras, more details on shielding can be found ![here](https://matrix.umd.edu/research/open-air-land-lab)

The Lab 2 is the location where the crazyradios constantly fail to communicate reliably which leads to crazyflies loosing poses, A detailed description of the issue can be found in the discussion links mentioned above. We specifically tested for Radio Latency between two Labs and the Lab 1 environment has better Latency diagnostics then Lab 2 which itself seems counter intuitive. If you have any solutions to these it would be welcome.

Thanks
Darshit

**Comment by whoenig** [comment]
@ataffanel and @knmcguire: This is what we briefly discussed today in our meeting as well, with focus on how we can test the various software stacks better for high packet loss environments.

The question here is also especially about the RF component: Could RF shielding of a room actually physically worsen the radio performance? If so, are there any countermeasures (e.g., would using an external antenna help)?

**Reply by ataffanel** [reply]
I am not sure what effect RF shielding can have. If is it not reflective it should actually help a lot, if there is reflections though this can create a lot of local attenuation of the signal by standing wave (ie. this is the reason microwave oven turns the food ....).

One thing to try would be to reduce the output power of the Crazyradio: this way if this is a problem of reflections, reducing power might help. I unfortunately do not have a lot of other idea, this is a typical case where channel hopping could help a lot as well (moving just 8MHz can move the standing wave local minimal away from the receiving antenna), but this is not something we have working in the Crazyradio right now. The principle could be tested by checking if BLE connection from the phone app to the Crazyflie (or any other bluetooth connection for that matter) is stable or not in this environment.

**Reply by darshit-desai** [reply]
Does a connection with a bluetooth mouse and a PC count as a valid test? And if so how would I test it?

Also, the two other methods of reducing the radio power and channel hopping can it be done by changing any code in the firmware of CF or crazyradio?

I am also not sure what type of shielding is there in the lab, I will confirm it soon

**Reply by ataffanel** [reply]
Yes connecting with a BLE mouse would maybe give some information. At least to see if the performance are similar in and out of the space (in terms of range for example)

As for the output power, it can be configured over USB but I am not sure if it is possible to tell Crazyswarm to set the radio power.

**Reply by whoenig** [reply]
There is no setting in Crazyswarm, but with a code change in https://github.com/bitcraze/crazyflie-link-cpp/blob/master/src/Crazyradio.cpp#L37, and recompiling it should work.

**Reply by darshit-desai** [reply]
So for the suggestion that @whoenig gave, I would change the value passed in the setPower function and recompile, in that case let's assume I figure out to change that code, isn't the firmware not supposed to be changed because it was modified for multi cast communication, I am using the firmware which was mentioned [here](https://github.com/USC-ACTLab/crazyswarm/tree/master/prebuilt)

For the suggestion that @ataffanel mentioned, if I use the crazyflie iOS app with a BLE bluetooth connection, the phone can only connect with one CF at a time, I can check out the performance on range but that would be it. Would it be a sufficient baseline for checking the bluetooth connection stability? I will get back to you of the type of shielding inside the lab walls soon.

**Reply by whoenig** [reply]
The code I pointed to is part of Crazyswarm2, not of any firmware. Changing that line followed by `colcon build` is all it takes.

Arnaud proposed using a bluetooth mouse, not a Crazyflie for testing.

**Reply by darshit-desai** [reply]
Ok that's my bad, I thought of firmware because of the bitcraze repo link, but I just noticed each crazyswarm directory has it's own firmware already in their directory tree.

Edit: @whoenig what do I need to change here? I need your help with that

About the bluetooth testing, I figured out that finding a mouse running on BLE might not be easier and it would be actually easier to just test it with a CF and the mobile app, that's what Arnaud had initially suggested.

**Reply by darshit-desai** [reply]
@whoenig Can you please help me for the change that I need to do for low power config of radios?
@ataffanel if there are any other parameters to check over BLE other then range please share it with me

**Reply by whoenig** [reply]
You change https://github.com/bitcraze/crazyflie-link-cpp/blob/master/src/Crazyradio.cpp#L37 (i.e., `setPower(Power_0DBM);`) to a different power value (defined in https://github.com/bitcraze/crazyflie-link-cpp/blob/master/src/Crazyradio.h#L27-L33), followed by `colcon build`.

**Reply by ataffanel** [reply]
I guess range and connection stability are the two main things to check over BLE. If you have the android client working or an interactive device like mouse/keyboard, roughly testing latency would also be great. Basically you need to check if the BLE link operates normally or not.

**Reply by darshit-desai** [reply]
Hi @whoenig I tried as you recommended above in Crazyswarm2, But since CSW2 already has issues with safelink because of which the Crazyflies are unable to fly in sync (I had to add unicasts which was told by Eric in a different issue). I was thinking of doing this change in crazyswarm 1, So for doing that change I would be changing in the same file inside the same file https://github.com/whoenig/crazyflie_cpp/blob/25bc72c120f8cea6664dd24e334eefeb7c9606ca/src/Crazyradio.cpp#L36 or in the ros_ws of crazyswarm at this path: ~/crazyswarm/ros_ws/src/crazyflie_tools/crazyflie_cpp/src/Crazyradio.cpp . I wanted to confirm this before changing it and testing it. Later I would need to catkin_make the ros_ws or use the bash file (build.sh) command which does the same thing

**Reply by whoenig** [reply]
That's correct. Although before switching to Crazyswarm(1), it would make sense to run `ros2 run crazyflie comCheck` with and without the Power change to see if it helps in your environment.

**Reply by darshit-desai** [reply]
Hi @whoenig @ataffanel ,

I have done the ros test using crazyswarm1 like last time, I made a bash file to run the comm check cpp file 10-20 times and extracted the stdout output in text files for each of the drones, It couldn't be run in parallel since it showed me libusb errors so mostly this was done using one drone at a time.

I have also used a Python script later to plot the relevant data points of the drones, the results are in the drive link below along with the raw txt files and the bash files I used to run the ros cpp script.

Note for some cases where even the script timed out, the values of that data point are set to -10.0, which indicates no data was collected.

Also, I'd like to point out that all the results shown here are of the lab environment where the potential of attenuation due to multiple reflections of radio signals or standing wave phenomena could be present. I am still entirely unsure of the cause of this since even with reducing the power levels, the radio pinging test results aren't encouraging.

The below is the link to the files I have uploaded on google drive:
https://drive.google.com/drive/folders/1_KkbCotpCIU0n4e-pog7KAxNLNFA-_cK?usp=sharing

The folder structure of where the results are present is shown below, For time efficiency I did some tests only for 24 drones, each test with a different radio power level has its results inside the folder,

```BASH
.
├── Power0Dbm
│   ├── 9 drones
│   │   ├── commcheckcli.txt
│   │   ├── comm_check.sh
│   │   ├── drone1_commcheck.txt
│   │   ├── drone2_commcheck.txt
│   │   ├── drone3_commcheck.txt
│   │   ├── drone4_commcheck.txt
│   │   ├── drone5_commcheck.txt
│   │   ├── drone6_commcheck.txt
│   │   ├── drone7_commcheck.txt
│   │   ├── drone8_commcheck.txt
│   │   └── drone9_commcheck.txt
│   ├── read_and_plot_USMSM.py
│   └── Results_Power0Dbm
│       └── 9 drones
├── PowerM12Dbm
│   ├── 24 drones
│   │   ├── commcheckcli.txt
│   │   ├── comm_check.sh
│   │   ├── drone10_commcheck.txt
│   │   ├── drone11_commcheck.txt
│   │   ├── drone12_commcheck.txt
│   │   ├── drone13_commcheck.txt
│   │   ├── drone14_commcheck.txt
│   │   ├── drone15_commcheck.txt
│   │   ├── drone16_commcheck.txt
│   │   ├── drone17_commcheck.txt
│   │   ├── drone18_commcheck.txt
│   │   ├── drone19_commcheck.txt
│   │   ├── drone1_commcheck.txt
│   │   ├── drone20_commcheck.txt
│   │   ├── drone21_commcheck.txt
│   │   ├── drone22_commcheck.txt
│   │   ├── drone23_commcheck.txt
│   │   ├── drone24_commcheck.txt
│   │   ├── drone2_commcheck.txt
│   │   ├── drone3_commcheck.txt
│   │   ├── drone4_commcheck.txt
│   │   ├── drone5_commcheck.txt
│   │   ├── drone6_commcheck.txt
│   │   ├── drone7_commcheck.txt
│   │   ├── drone8_commcheck.txt
│   │   └── drone9_commcheck.txt
│   ├── read_and_plot_USMSM.py
│   └── Results_PowerM12Dbm
│       └── 24 drones
├── PowerM18Dbm
│   ├── 24 drones
│   │   ├── commcheckcli.txt
│   │   ├── comm_check.sh
│   │   ├── drone10_commcheck.txt
│   │   ├── drone11_commcheck.txt
│   │   ├── drone12_commcheck.txt
│   │   ├── drone13_commcheck.txt
│   │   ├── drone14_commcheck.txt
│   │   ├── drone15_commcheck.txt
│   │   ├── drone16_commcheck.txt
│   │   ├── drone17_commcheck.txt
│   │   ├── drone18_commcheck.txt
│   │   ├── drone19_commcheck.txt
│   │   ├── drone1_commcheck.txt
│   │   ├── drone20_commcheck.txt
│   │   ├── drone21_commcheck.txt
│   │   ├── drone22_commcheck.txt
│   │   ├── drone23_commcheck.txt
│   │   ├── drone24_commcheck.txt
│   │   ├── drone2_commcheck.txt
│   │   ├── drone3_commcheck.txt
│   │   ├── drone4_commcheck.txt
│   │   ├── drone5_commcheck.txt
│   │   ├── drone6_commcheck.txt
│   │   ├── drone7_commcheck.txt
│   │   ├── drone8_commcheck.txt
│   │   └── drone9_commcheck.txt
│   ├── read_and_plot_USMSM.py
│   └── Results_PowerM18Dbm
│       └── 24 drones
└── PowerM6Dbm
    ├── 24 Drones
    │   ├── commcheckcli.txt
    │   ├── comm_check.sh
    │   ├── drone10_commcheck.txt
    │   ├── drone11_commcheck.txt
    │   ├── drone12_commcheck.txt
    │   ├── drone13_commcheck.txt
    │   ├── drone14_commcheck.txt
    │   ├── drone15_commcheck.txt
    │   ├── drone16_commcheck.txt
    │   ├── drone17_commcheck.txt
    │   ├── drone18_commcheck.txt
    │   ├── drone19_commcheck.txt
    │   ├── drone1_commcheck.txt
    │   ├── drone20_commcheck.txt
    │   ├── drone21_commcheck.txt
    │   ├── drone22_commcheck.txt
    │   ├── drone23_commcheck.txt
    │   ├── drone24_commcheck.txt
    │   ├── drone2_commcheck.txt
    │   ├── drone3_commcheck.txt
    │   ├── drone4_commcheck.txt
    │   ├── drone5_commcheck.txt
    │   ├── drone6_commcheck.txt
    │   ├── drone7_commcheck.txt
    │   ├── drone8_commcheck.txt
    │   ├── drone9_commcheck.txt
    │   └── obsolete
    ├── 9 Drones
    │   ├── commcheckcli.txt
    │   ├── comm_check.sh
    │   ├── drone1_commcheck.txt
    │   ├── drone2_commcheck.txt
    │   ├── drone3_commcheck.txt
    │   ├── drone4_commcheck.txt
    │   ├── drone5_commcheck.txt
    │   ├── drone6_commcheck.txt
    │   ├── drone7_commcheck.txt
    │   ├── drone8_commcheck.txt
    │   ├── drone9_commcheck.txt
    │   └── obsolete
    ├── read_and_plot_USMSM.py
    └── Results_PowerM6Dbm
        ├── 24 Drones
        └── 9 Drones

20 directories, 104 files
```

**Reply by knmcguire** [reply]
Hi! Thanks for sharing the files. I just want to let you know that currently we are working on some deadlines and we won't be able to dig into the crazyradio2 issue just yet. We hope that in some weeks us and arnaud can sit down and figure it out

**Reply by whoenig** [reply]
This is Crazyradio PA, not Crazyradio2, although I would not expect that Crazyradio2 shows any difference.

I did look at the data, but I am as puzzled as before and I don't know enough about RF wave propagation to interpret the results. It might be worth a try to put an external antenna on the very badly performing drones and see if that helps. The Bolt comes with a matching antenna, but perhaps Bitcraze can also tell you where/what antennas to buy.

**Reply by darshit-desai** [reply]
I also had the same observation, technically a problem of standing wave or multiple reflections should be solved by either reducing the radio power or by using frequency hopping. The first one we tried already but some guys who have faced such radio issues in domains other then drones said that a -12 dBm reduction in power is still less, if the problem is about reflections the reductions should be on the range of -60 dBm. Also for frequency hopping I assume the firmware is taking care of it internally since in the crazyradio.cpp there was no mention of it.

**Reply by whoenig** [reply]
The Crazyradio PA chip only supports down to -18dBm. Frequency hoping is not used at all and part of the URI is the fixed frequency. If you use Bluetooth, there is frequency hoping, but not sure if/how it is usable from a PC and how many drones could be controlled.
